var/spack/repos/builtin/packages/doxygen/package.py

```diff
@@ -33,6 +33,30 @@ class Doxygen(CMakePackage):
     variant('mscgen', default=False,
             description='Build with support for code graphs from mscgen.')
 
+    executables = ['doxygen']
+
+    @classmethod
+    def determine_version(cls, exe):
+        output = Executable(exe)('-v', output=str, error=str)
+        return output
+
+    @classmethod
+    def determine_variants(cls, exes, version_str):
+        variants = ''
+        try:
+            Executable("dot")('-V', output=str, error=str)
+            variants += "+graphviz"
+        except ProcessError:
+            variants += "~graphviz"
```

Comment:
(question not request) Since we don't use the output, would it be better to use `which` instead of calling the executable?

Reply:
I actually kind of feel weird about this checking and I hesitated even put it in.  `dot` and `mscgen` are not provided by the `doxygen` package on any distribution that I've ever used;  Rather they are run time dependencies used by doxygen to generate figures such as class/interaction diagrams and sequence diagrams respectively.

It feels like it should be possible to say please use the external doxygen, but please build graphviz and mscgen for me, but I was unsure how to do that.  However if I leave off the variant checking then the most common use case for doxygen: `doxygen+graphviz` doesn't work with `spack external find doxygen` because it guesses `doxygen~graphviz~mscgen`.  

How should I address this?

Reply:
What you say makes a lot of sense. The issue is that the current model for external detection is "static" (we inspect the filesystem or the output of an executable at a certain point in time and we detect which spec is installed). We should probably generalize it in the sense you say when / if we decide not to truncate DAG nodes on external detection.

That said, I think the way you deal with it is fine given the current model. The suggestion was just to avoid calling the executable if we don't check the output.